### PR TITLE
SyncLabs V2

### LINKED ISSUE #1196: Sync Labs: Add dialog box when nothing is selected
It should act similar to when ColorsLab has no shape selected, for both Copy and Paste

### PR DESCRIPTION
**Pull request by Kenneth-LJS**
Fixes #1196

Change Log:

1. List items with no images have their image box automatically hidden.
Before:
image

After:
image

2. Helpful error messages when no items are selected during copy/paste.
image
image

3. New formats copied will automatically be selected.
image

4. The selected format will show the name of the original shape
Before:
image

After:
image

5. Effects being pasted will apply on ALL shapes in selection.

6. Formats will not appear when not applicable. E.g. Options for arrowhead will not show when copying a rectangle's format.

7. Display images have been added to the various formats.
image

8. Unit tests have been added to ensure that all new formats added by developers have implemented the required functionality.

### CODE REVIEW

File: PowerPointLabs/PowerPointLabs/SyncLab/ObjectFormats/FillFormat.cs [hunk 1]
```diff
@@ -1,5 +1,6 @@
 ﻿using System.Drawing;
 using Microsoft.Office.Interop.PowerPoint;
+using Graphics = PowerPointLabs.Utils.Graphics;
```
Comment (initialshl) [inline PowerPointLabs/PowerPointLabs/SyncLab/ObjectFormats/FillFormat.cs]: Would be good to separate each import group with line breaks:
```
using System.Drawing;

using Microsoft.Office.Interop.PowerPoint;

using Graphics = PowerPointLabs.Utils.Graphics;

File: PowerPointLabs/PowerPointLabs/SyncLab/ObjectFormats/LineArrowFormat.cs [hunk 2]
```diff
@@ -1,12 +1,22 @@
...
 {
     class LineArrowFormat
     {
         public static bool CanCopy(Shape formatShape)
         {
+            try
+            {
+                SyncFormat(formatShape, formatShape);
+            }
+            catch (Exception)
```
Comment (initialshl) [inline PowerPointLabs/PowerPointLabs/SyncLab/ObjectFormats/LineArrowFormat.cs]: Maybe log the exception

File: PowerPointLabs/PowerPointLabs/SyncLab/View/SyncFormatDialog.xaml.cs [hunk 3]
```diff
@@ -96,6 +94,79 @@ private void OkButton_Click(object sender, RoutedEventArgs e)
...
+        private void ScrollToTop()
+        {
+            if (treeView.Items.IsEmpty)
+            {
+                treeView.Items.MoveCurrentToFirst();
+                //(treeView.Items[0] as TreeViewItem).BringIntoView();
+            } 
+            
+            /*
+            ScrollViewer scrollViewer = GetScrollViewer(treeView);
```
Comment (initialshl) [inline PowerPointLabs/PowerPointLabs/SyncLab/View/SyncFormatDialog.xaml.cs]: Remove unused code if possible?

File: PowerPointLabs/PowerPointLabs/SyncLab/View/SyncFormatListItem.xaml.cs [hunk 4]
```diff
@@ -50,12 +51,30 @@ public Bitmap Image
             set
             {
                 image = value;
+                UpdateImage();
+            }
+        }
+
+        private void UpdateImage()
+        { // if image isn't set, fill the control with the text box
```
Comment (initialshl) [inline PowerPointLabs/PowerPointLabs/SyncLab/View/SyncFormatListItem.xaml.cs]: Put the comment on a new line?

File: PowerPointLabs/Test/UnitTest/SyncLab/SyncLabFormatTest.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                AssertValidFormat(node);
+            }
+        }
+
+        private void AssertValidFormat(FormatTreeNode node)
+        {
+            if (node.IsFormatNode)
+            {
+                AssertValidFormatType(node.Format.FormatType);
+            } else
```
Comment (initialshl) [inline PowerPointLabs/Test/UnitTest/SyncLab/SyncLabFormatTest.cs]: Missing a line break

File: PowerPointLabs/Test/UnitTest/SyncLab/SyncLabFormatTest.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
+﻿using System;
+
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using PowerPointLabs.SyncLab;
+using PowerPointLabs.SyncLab.ObjectFormats;
+using System.Reflection;
```
Comment (initialshl) [inline PowerPointLabs/Test/UnitTest/SyncLab/SyncLabFormatTest.cs]: Group the imports in alphabetical order

File: PowerPointLabs/PowerPointLabs/SyncLab/View/SyncFormatListItem.xaml.cs [hunk 7]
```diff
@@ -50,12 +51,30 @@ public Bitmap Image
...
+            {
+                imageBox.Visibility = Visibility.Hidden;
+                label.Margin = new Thickness(30, label.Margin.Top,
+                            label.Margin.Right, label.Margin.Bottom);
+                return;
+            }
+            else
+            {
                 BitmapSource source = Imaging.CreateBitmapSourceFromHBitmap(
-                        image.GetHbitmap(),
```
Comment (initialshl) [inline PowerPointLabs/PowerPointLabs/SyncLab/View/SyncFormatListItem.xaml.cs]: Better to keep the indentations here?

File: PowerPointLabs/PowerPointLabs/SyncLab/ObjectFormats/LineArrowFormat.cs (hunk 2 above)
Comment (Kenneth-LJS) [inline PowerPointLabs/PowerPointLabs/SyncLab/ObjectFormats/LineArrowFormat.cs]: This isn't meant to be an actual exception. There does not seem to be a way to check if a certain shape actually has an arrowhead format, than to access it. However if we access it and it isn't there, it throws an exception. This is why I am catching it here, then the function returns false, that there is no arrow :P

Comment (initialshl) [inline PowerPointLabs/PowerPointLabs/SyncLab/ObjectFormats/LineArrowFormat.cs]: Ohh, maybe could add a comment there with the reason

**Pull request merged**